File: python/86_分隔链表_medium.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        if not head:
            return head
        dummyHead, tmp = ListNode(0, head), ListNode(0)
        p, pt = dummyHead, tmp
        while p.next:
            if p.next.val < x:
                p = p.next
            else:
                tmp.next = p.next
                p.next = p.next.next
                tmp = tmp.next
                tmp.next = None
        p.next = pt.next
        return dummyHead.next

File: cpp/86_分隔链表_medium.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* partition(ListNode* head, int x) {
        if(!head)
            return head;
        ListNode *tmp = new ListNode(0);
        ListNode *dummyHead = new ListNode(0, head);
        ListNode *p = dummyHead;
        ListNode *pt = tmp;
        while(p->next != nullptr){
            if(p->next->val < x)
                p = p->next;
            else{
                tmp->next = p->next;
                p->next = p->next->next;
                tmp = tmp->next;
                tmp->next = nullptr;
            }
        }
        p->next = pt->next;
        return dummyHead->next;
        
    }
};